Sources/SwiftCompilerPluginMessageHandling/JSON/JSONDecoding.swift

```diff
@@ -0,0 +1,1205 @@
+ ```
+
+ will be scanned by 'JSONScanner' into a map like:
+
+ ```
+ <OM> == Object Marker
+ <AM> == Array Marker
+ <SS> == Simple String (a variant of String that can has no escapes and can be passed directly to a UTF-8 parser)
+ <NM> == Number Marker
+ <TL> == NULL Marker
+ map: [
+   0: <OM>,      -- object marker
+   1: 15,        |  `- number of *map* elements in this collection
+   2: <SS>,      | --- key 1: 'array'
+   3: <int_ptr>, | |   |- pointer in the payload
+   4: 5,         | |   `- length
+   5: <AM>,      | --- value 1: array
+   6: 4,         | |   `- number of *map* elements in the array
+   7: <NM>,      | | -- arr elm 1: '-1.3'
+   8: <int_ptr>, | | |
+   9: 4,         | | |
+  10: <TL>,      | | -- arr elm 2: 'true'
+  11: <SS>,      | --- key 2: 'number'
+  12: <int_ptr>, | |
+  13: 6,         | |
+  14: <NM>       | --- value 2: '42'
+  15: <int_ptr>, | |
+  16: 2,         | |
+ ]
+ ```
+ To decode '<root>.number' value:
+ 1. Index 0 indicates it's a object.
+ 2. Parse a key string at index 2, which is not a  match for "number"
+ 3. Skip the key's value by finding it's an array, then its 'index(afterValue:)' which is 11
+ 4. Parse a key string at index 11, matching "number"
+ 5. Parse a value number at the pointer of index 15, length at index 16
+*/
+
+private struct JSONMap {
+  enum Descriptor: Int {
+    case nullKeyword  // [desc]
+    case trueKeyword  // [desc]
+    case falseKeyword  // [desc]
+    case number  // [desc, pointer, length]
+    case simpleString  // [desc, pointer, length]
+    case string  // [desc, pointer, length]
+    case object  // [desc, count, (key, value)...]
+    case array  // [desc, count, element...]
+  }
+  let data: [Int]
+
+  /// Top-level value.
+  func withValue<T>(_ fn: (JSONMapValue) throws -> T) rethrows -> T {
+    try withExtendedLifetime(data) {
+      try data.withUnsafeBufferPointer { buf in
+        try fn(JSONMapValue(data: buf.baseAddress!))
+      }
+    }
+  }
+}
+
+private struct JSONMapBuilder {
+  var mapData: [Int]
+
+  init() {
+    mapData = []
+    mapData.reserveCapacity(128)  // 128 is good enough for most PluginMessage.
+  }
+
+  @inline(__always)
+  mutating func record(_ descriptor: JSONMap.Descriptor) {
+    mapData.append(descriptor.rawValue)
+  }
+
+  @inline(__always)
+  mutating func record(_ descriptor: JSONMap.Descriptor, range: Range<UnsafePointer<UInt8>>) {
+    mapData.append(descriptor.rawValue)
+    mapData.append(Int(bitPattern: range.lowerBound))
+    mapData.append(range.count)
+  }
+
+  @inline(__always)
+  mutating func startCollection(_ descriptor: JSONMap.Descriptor) -> Int {
+    let handle = mapData.count
+    mapData.append(descriptor.rawValue)
+    mapData.append(0)  // Count, this will be updated in closeCollection()
+    return handle
+  }
+
+  @inline(__always)
+  mutating func closeCollection(handle: Int) {
+    // 'handle': descriptor index.
+    // 'handle+1': counter index.
+    mapData[handle + 1] = mapData.count - handle - 2
```

Comment:
There's no reason to `- 2` here. Only usage (i.e. `JSONMapValue.mapSize`) is doing `+ 2`. Let's just store the `mapSize` value, so `mapSize` can return the stored value as-is.